### PR TITLE
DM-37955: Update measureApCorrTask to use robust median outlier rejection.

### PR DESCRIPTION
**Pull request by erykoff**
This replaces the default source selector to use all possible (unflagged) stars, and additional checks on the fit degree after outlier rejection.

### CODE REVIEW

File: python/lsst/meas/algorithms/measureApCorr.py [hunk 1]
```diff
@@ -34,32 +35,39 @@
...
-                read: {name}_instFlux, {name}_instFluxErr, {name}_flag
-                added: apcorr_{name}_used
-        """
-        self.flux = schema.find(name + "_instFlux").key
-        self.err = schema.find(name + "_instFluxErr").key
-        self.flag = schema.find(name + "_flag").key
-        self.used = schema.addField("apcorr_" + name + "_used", type="Flag",
-                                    doc="set if source was used in measuring aperture correction")
+        self.fluxName = name + "_instFlux"
+        if not schema.extract(self.fluxName):
```
Comment (TallJimbo) [inline python/lsst/meas/algorithms/measureApCorr.py:56]: `if self.fluxName not in schema:` is more idiomatic, and I sure hope it's not slower.

File: python/lsst/meas/algorithms/measureApCorr.py [hunk 2]
```diff
@@ -164,103 +195,126 @@ def run(self, exposure, catalog):
...
-                apCorrData = apCorrData[keep]
-                indices = indices[keep]
 
-            # Final fit after clipping
-            apCorrField = ChebyshevBoundedField.fit(bbox, x, y, apCorrData, ctrl)
+            allBad = False
+            for iteration in range(self.config.numIter):
+                resid = z - fitValues
+                # We allow floating-point slop to allow for field estimation
+                # noise in a completely flat field.
```
Comment (TallJimbo) [inline python/lsst/meas/algorithms/measureApCorr.py:253]: Does this avoid a divide-by-zero somewhere?  I think that means a slightly longer comment here would be useful.

File: tests/test_measureApCorr.py [hunk 3]
```diff
@@ -199,6 +207,49 @@ def testApertureMeasTens(self):
...
+        self.meas_apCorr_task.config.minDegreesOfFreedom = 4
+
+        for name in self.names:
+            nameAp = name + self.apNameStr
+            sourceCat[nameAp + "_instFlux"][0] = 100.0
+
+        with self.assertRaisesRegex(RuntimeError, "only 4 sources remain"):
+            self.meas_apCorr_task.run(catalog=sourceCat, exposure=self.exposure)
+
+        # With the measurement algorithm declared as something that might fail, should not get an exception
```
Comment (TallJimbo) [inline tests/test_measureApCorr.py:248]: Something in this sentence seems awry.

**Review by TallJimbo — APPROVED** [review]

File: python/lsst/meas/algorithms/measureApCorr.py (hunk 2 above)
Comment (erykoff) [reply python/lsst/meas/algorithms/measureApCorr.py:253]: I don't think so, but I've rephrased the comment to be clear this shows up in tests when the residual field is completely flat and median_abs_deviation doesn't return identically zero in that case.

File: tests/test_measureApCorr.py (hunk 3 above)
Comment (erykoff) [reply tests/test_measureApCorr.py:248]: I copied it from another test, but I have rephrased both of them to be clearer.

**Pull request merged**